File: homing-workspace/src/main/java/hue/captains/singapura/js/homing/workspace/party/LayoutSecretaryModule.java
package hue.captains.singapura.js.homing.workspace.party;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for {@code LayoutParty} — extracted as a standalone JS
 * module per the Diligent Secretaries doctrine's third pillar: behaviors
 * must be unit-testable as pure functions, which requires being
 * importable in isolation (rather than inlined in the chrome's bodyJs).
 *
 * <p>Exports a single {@code LayoutSecretary} JS object with two members:
 * {@code initial} (the starting state record) and {@code behavior} (the
 * pure (state, envelope) → Step function). The {@code WorkspaceLayout}
 * chrome imports both and passes them to the Party constructor.</p>
 *
 * <h2>State shape (Diligent Secretaries enrichment)</h2>
 *
 * <pre>{@code
 * {
 *     fullscreen    : boolean,            // current value
 *     lastChangedBy : AgentId | null,     // provenance of most recent change
 *     recentUnknown : [{ kind, from }]    // last N unrecognised messages (bounded)
 * }
 * }</pre>
 *
 * <h2>Message kinds</h2>
 *
 * <ul>
 *   <li>{@code FullscreenToggleRequested} — flip and broadcast {@code FullscreenChanged}.</li>
 *   <li>{@code FullscreenSetRequested(on)} — set if different; broadcast on change; no-op if same.</li>
 *   <li>Anything else — append to {@code recentUnknown} (bounded at 10); no action emitted.</li>
 * </ul>
 *
 * @since RFC 0028 cycle 6 phase 1 — Secretary-as-module refactor for JS test harness
 */
public record LayoutSecretaryModule() implements DomModule<LayoutSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior} members. */
    public record LayoutSecretary() implements Exportable._Constant<LayoutSecretaryModule> {}

    public static final LayoutSecretaryModule INSTANCE = new LayoutSecretaryModule();

    @Override
    public ImportsFor<LayoutSecretaryModule> imports() {
        return ImportsFor.<LayoutSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<LayoutSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new LayoutSecretary()));
    }
}

File: homing-workspace/src/main/resources/homing/js/hue/captains/singapura/js/homing/workspace/party/LayoutSecretaryModule.js
// =============================================================================
// LayoutSecretaryModule — the Secretary for the framework's LayoutParty.
//
// Per the Diligent Secretaries doctrine: pure (state, envelope) -> Step
// behavior plus initial state, both exposed as members of a single
// top-level object for clean framework import + isolated test harness load.
//
// Diligence enrichment in state:
//   - fullscreen     : current value
//   - lastChangedBy  : provenance of the most recent state change (envelope.from)
//   - recentUnknown  : last 10 unrecognised messages (kind + origin), bounded
//
// Side-effect-free; no DOM, no console, no captures. Per the
// Lambdas-Are-Functional-Object-Shortcuts doctrine, this entire module
// is a Functional Object simplification — pure behavior + immutable state shape.
// =============================================================================

var LayoutSecretary = {

    initial: {
        fullscreen   : false,
        lastChangedBy: null,
        recentUnknown: []
    },

    /**
     * Pure routing function. (state, envelope) -> { newState, actions }.
     * Same inputs always produce same outputs; no side effects.
     */
    behavior: function (state, envelope) {
        var msg = envelope.message;

        switch (msg.kind) {

            case "FullscreenToggleRequested": {
                var next = !state.fullscreen;
                return {
                    newState: {
                        fullscreen   : next,
                        lastChangedBy: envelope.from,
                        recentUnknown: state.recentUnknown
                    },
                    actions: [{
                        kind   : "BroadcastToMembers",
                        message: { kind: "FullscreenChanged", on: next }
                    }]
                };
            }

            case "FullscreenSetRequested": {
                // No-op when already at the requested value; preserves state identity
                // so dispatch is exactly as expensive as the no-op branch.
                if (!!msg.on === state.fullscreen) {
                    return { newState: state, actions: [] };
                }
                return {
                    newState: {
                        fullscreen   : !!msg.on,
                        lastChangedBy: envelope.from,
                        recentUnknown: state.recentUnknown
                    },
                    actions: [{
                        kind   : "BroadcastToMembers",
                        message: { kind: "FullscreenChanged", on: !!msg.on }
                    }]
                };
            }

            default: {
                // Diligent defensive observability: track unknowns in state so
                // they surface via inspect() rather than silently disappearing.
                // Bounded at 10 entries via .slice(-10) so state cannot grow.
                var recent = state.recentUnknown.concat([{
                    kind: msg.kind,
                    from: envelope.from
                }]).slice(-10);
                return {
                    newState: {
                        fullscreen   : state.fullscreen,
                        lastChangedBy: state.lastChangedBy,
                        recentUnknown: recent
                    },
                    actions: []
                };
            }
        }
    }
};
